### PR TITLE
Make DbKvs much faster for certain types of queries

### PR DESCRIPTION
**Pull request by j-baker**
This code is in a very unfinished state, but I wanted to send this out
to see what the response is like.

At the moment when we do a read from Atlas, we do a
query which is linear in the number of revisions of a row (have to scan over all column versions).

This is sad and unnecessary when we have a column selection, where we can just do reverse index lookups for each one on the primary key.

Instead, use a lateral join, turning this into a reverse index scan. I'm not a postgres pro - I just spent a little while hacking away at this. The result is much faster for certain queries.

Some query plans:

On a cell with 100 versions:

```
atlas=# EXPLAIN ANALYSE SELECT wrap.row_name, wrap.col_name, wrap.ts,
wrap.val
        FROM table wrap, (
        /* GET_LATEST_ROW_INNER (table) */
             SELECT m.row_name, m.col_name, max(m.ts) as ts
             FROM gatekeeper_v2__resources m
             WHERE m.row_name =  '\x8000000000000064'
             AND m.ts < 123611
             GROUP BY m.row_name, m.col_name ) i
        WHERE wrap.row_name = i.row_name AND wrap.col_name = i.col_name
AND wrap.ts = i.ts;
                                                                                 QUERY
PLAN

-------------------------------------------------------------------------------------------------------------------------------
---------------------------------------------
 Nested Loop  (cost=0.83..8.92 rows=1 width=85) (actual
time=0.162..0.164 rows=1 loops=1)
   ->  GroupAggregate  (cost=0.41..4.45 rows=1 width=19) (actual
time=0.155..0.155 rows=1 loops=1)
         ->  Index Only Scan using pk_gatekeeper_v2__resources on
gatekeeper_v2__resources m  (cost=0.41..4.43 rows=1 width=19)
 (actual time=0.031..0.124 rows=102 loops=1)
               Index Cond: ((row_name = '\x8000000000000064'::bytea) AND
(ts < 123611))
               Heap Fetches: 102
   ->  Index Scan using pk_table on
table wrap  (cost=0.41..4.45 rows=1 width=85) (actual
 time=0.005..0.005 rows=1 loops=1)
         Index Cond: ((row_name = m.row_name) AND (col_name =
m.col_name) AND (ts = (max(m.ts))))
 Total runtime: 0.208 ms
(8 rows)

atlas=# EXPLAIN ANALYSE SELECT cells.row_name, cells.col_name, data.ts,
data.val
atlas-# FROM (SELECT t1.row_name AS row_name, t2.col_name AS col_name
atlas(#       FROM (SELECT unnest(array[decode('8000000000000064',
'hex')]) AS row_name) t1, (SELECT unnest(array[decode('70', 'hex')]) AS
col_name) t2) cells
atlas-# INNER JOIN LATERAL (SELECT ts, val FROM table
tbl
atlas(# WHERE tbl.row_name = cells.row_name
atlas(#     AND tbl.col_name = cells.col_name
atlas(#     AND tbl.ts < 123611
atlas(#     ORDER BY tbl.ts
atlas(#     DESC LIMIT 1) data
atlas-# ON true;
                                                                                      QUERY
PLAN

-------------------------------------------------------------------------------------------------------------------------------
-------------------------------------------------------
 Nested Loop  (cost=0.41..44827.26 rows=10000 width=138) (actual
time=0.038..0.040 rows=1 loops=1)
   ->  Result  (cost=0.00..0.51 rows=100 width=0) (actual
time=0.003..0.003 rows=1 loops=1)
   ->  Nested Loop  (cost=0.41..447.26 rows=100 width=106) (actual
time=0.032..0.034 rows=1 loops=1)
         ->  Result  (cost=0.00..0.51 rows=100 width=0) (actual
time=0.001..0.001 rows=1 loops=1)
         ->  Limit  (cost=0.41..4.44 rows=1 width=74) (actual
time=0.029..0.030 rows=1 loops=1)
               ->  Index Scan Backward using pk_table
on table tbl  (cost=0.41..4.44 rows
=1 width=74) (actual time=0.028..0.028 rows=1 loops=1)
                     Index Cond: ((row_name =
(unnest('{"\\x8000000000000064"}'::bytea[]))) AND (col_name =
(unnest('{"\\x70"}'
::bytea[]))) AND (ts < 123611))
 Total runtime: 0.066 ms
(8 rows)
```

and on a row with 40,000 old versions

```
atlas=# EXPLAIN ANALYSE SELECT wrap.row_name, wrap.col_name, wrap.ts,
wrap.val
atlas-#         FROM table wrap, (
atlas(#         /* GET_LATEST_ROW_INNER (table) */
atlas(#              SELECT m.row_name, m.col_name, max(m.ts) as ts
atlas(#              FROM table m
atlas(#              WHERE m.row_name =  '\x73657175656e6365'
atlas(#              AND m.ts < 123611
atlas(#              GROUP BY m.row_name, m.col_name ) i
atlas-#         WHERE wrap.row_name = i.row_name AND wrap.col_name =
i.col_name AND wrap.ts = i.ts;
                                                                           QUERY
PLAN
----------------------------------------------------------------------------------------------------------------------------------------------------------------
 Nested Loop  (cost=1619.69..1623.75 rows=1 width=56) (actual
time=25.656..25.674 rows=3 loops=1)
   ->  HashAggregate  (cost=1619.27..1619.28 rows=1 width=19) (actual
time=25.634..25.634 rows=3 loops=1)
         ->  Seq Scan on table m  (cost=0.00..1327.55
rows=38896 width=19) (actual time=0.015..12.465 rows=37663 loops=1)
               Filter: ((ts < 123611) AND (row_name =
'\x73657175656e6365'::bytea))
               Rows Removed by Filter: 10176
   ->  Index Scan using pk_table on
gatekeeper_v2__sequence wrap  (cost=0.41..4.45 rows=1 width=56) (actual
time=0.009..0.009 rows=1 loops=3)
         Index Cond: ((row_name = m.row_name) AND (col_name =
m.col_name) AND (ts = (max(m.ts))))
 Total runtime: 25.728 ms
(8 rows)

vs

atlas=# EXPLAIN ANALYSE SELECT cells.row_name, cells.col_name, data.ts,
data.val
atlas-# FROM (SELECT t1.row_name AS row_name, t2.col_name AS col_name
atlas(#       FROM (SELECT unnest(array[decode('73657175656e6365',
'hex')]) AS row_name) t1, (SELECT unnest(array[decode('74', 'hex')]) AS
col_name) t2) cells
atlas-# INNER JOIN LATERAL (SELECT ts, val FROM table
tbl
atlas(# WHERE tbl.row_name = cells.row_name
atlas(#     AND tbl.col_name = cells.col_name
atlas(#     AND tbl.ts < 123611
atlas(#     ORDER BY tbl.ts
atlas(#     DESC LIMIT 1) data
atlas-# ON true;
                                                                                        QUERY
PLAN

-----------------------------------------------------------------------------------------------------------------------------------------------------------------------
--------------------
 Nested Loop  (cost=0.41..5137.14 rows=10000 width=109) (actual
time=0.039..0.041 rows=1 loops=1)
   ->  Result  (cost=0.00..0.51 rows=100 width=0) (actual
time=0.006..0.006 rows=1 loops=1)
   ->  Nested Loop  (cost=0.41..50.36 rows=100 width=77) (actual
time=0.031..0.032 rows=1 loops=1)
         ->  Result  (cost=0.00..0.51 rows=100 width=0) (actual
time=0.002..0.003 rows=1 loops=1)
         ->  Limit  (cost=0.41..0.47 rows=1 width=45) (actual
time=0.028..0.028 rows=1 loops=1)
               ->  Index Scan Backward using pk_table
on table tbl  (cost=0.41..2080.89 rows=38896 width=45)
(actual time=0.026..0.
026 rows=1 loops=1)
                     Index Cond: ((row_name =
(unnest('{"\\x73657175656e6365"}'::bytea[]))) AND (col_name =
(unnest('{"\\x74"}'::bytea[]))) AND (ts < 123611))
 Total runtime: 0.082 ms
(8 rows)
```

---
This change is [Reviewable](https://reviewable.io/reviews/palantir/atlasdb/690)

### CODE REVIEW

File: atlasdb-dbkvs/src/main/java/com/palantir/atlasdb/keyvalue/dbkvs/impl/postgres/PostgresQueryFactory.java [hunk 1]
```diff
@@ -45,15 +45,35 @@ public FullQuery getLatestRowQuery(byte[] row,
                                        long ts,
                                        ColumnSelection columns,
                                        boolean includeValue) {
+        if (columns.allColumnsSelected() || columns.getSelectedColumns().size() > 50) {
```
Comment (j-baker) [inline atlasdb-dbkvs/src/main/java/com/palantir/atlasdb/keyvalue/dbkvs/impl/postgres/PostgresQueryFactory.java]: the 50 is just some random crap I wrote, dunno if anyone ever requests that much.

**Comment by j-baker** [comment]
Oh, and on a row with only one version:

```

atlas=# EXPLAIN ANALYSE SELECT wrap.row_name, wrap.col_name, wrap.ts, wrap.val
        FROM gatekeeper_v2__resources wrap, (
        /* GET_LATEST_ROW_INNER (gatekeeper_v2__resources) */
             SELECT m.row_name, m.col_name, max(m.ts) as ts
             FROM gatekeeper_v2__resources m
             WHERE m.row_name =  '\x800000000000bec3'
             AND m.ts < 160000
             GROUP BY m.row_name, m.col_name ) i
        WHERE wrap.row_name = i.row_name AND wrap.col_name = i.col_name AND wrap.ts = i.ts;
                                                                                QUERY PLAN

-----------------------------------------------------------------------------------------------------------------------------------------------------------------------
---
 Nested Loop  (cost=0.83..8.92 rows=1 width=85) (actual time=0.022..0.023 rows=1 loops=1)
   ->  GroupAggregate  (cost=0.41..4.45 rows=1 width=19) (actual time=0.018..0.018 rows=1 loops=1)
         ->  Index Only Scan using pk_gatekeeper_v2__resources on gatekeeper_v2__resources m  (cost=0.41..4.43 rows=1 width=19) (actual time=0.014..0.015 rows=1 loops=
1)
               Index Cond: ((row_name = '\x800000000000bec3'::bytea) AND (ts < 160000))
               Heap Fetches: 1
   ->  Index Scan using pk_gatekeeper_v2__resources on gatekeeper_v2__resources wrap  (cost=0.41..4.45 rows=1 width=85) (actual time=0.003..0.003 rows=1 loops=1)
         Index Cond: ((row_name = m.row_name) AND (col_name = m.col_name) AND (ts = (max(m.ts))))
 Total runtime: 0.057 ms
(8 rows)

atlas=# EXPLAIN ANALYSE SELECT cells.row_name, cells.col_name, data.ts, data.val
FROM (SELECT t1.row_name AS row_name, t2.col_name AS col_name
      FROM (SELECT unnest(array[decode('800000000000bec3', 'hex')]) AS row_name) t1, (SELECT unnest(array[decode('70', 'hex')]) AS col_name) t2) cells
INNER JOIN LATERAL (SELECT ts, val FROM gatekeeper_v2__resources tbl
WHERE tbl.row_name = cells.row_name
    AND tbl.col_name = cells.col_name
    AND tbl.ts < 160000
    ORDER BY tbl.ts
    DESC LIMIT 1) data
ON true;
                                                                                      QUERY PLAN

-----------------------------------------------------------------------------------------------------------------------------------------------------------------------
---------------
 Nested Loop  (cost=0.41..44827.26 rows=10000 width=138) (actual time=0.023..0.025 rows=1 loops=1)
   ->  Result  (cost=0.00..0.51 rows=100 width=0) (actual time=0.003..0.003 rows=1 loops=1)
   ->  Nested Loop  (cost=0.41..447.26 rows=100 width=106) (actual time=0.017..0.019 rows=1 loops=1)
         ->  Result  (cost=0.00..0.51 rows=100 width=0) (actual time=0.001..0.002 rows=1 loops=1)
         ->  Limit  (cost=0.41..4.44 rows=1 width=74) (actual time=0.014..0.014 rows=1 loops=1)
               ->  Index Scan Backward using pk_gatekeeper_v2__resources on gatekeeper_v2__resources tbl  (cost=0.41..4.44 rows=1 width=74) (actual time=0.013..0.013 r
ows=1 loops=1)
                     Index Cond: ((row_name = (unnest('{"\\x800000000000bec3"}'::bytea[]))) AND (col_name = (unnest('{"\\x70"}'::bytea[]))) AND (ts < 160000))
 Total runtime: 0.048 ms
```

**Comment by j-baker** [comment]
to be clear, code is intentionally low quality (this is very much a POC).

**Comment by joelea** [comment]
Closing this as it's not going to be looked at in the immediate future, however we are explicitly tracking it in #786

**Pull request closed without merging**